File: 141LinkedListCycle.cpp
//1. Use a fast pointer and a slow pointer. If they meet. It has cycle. O(N).
class Solution {
public:
    bool hasCycle(ListNode *head) {
        ListNode *slow = head;
        ListNode *fast = head;
        while(fast && fast->next){
            fast = fast -> next -> next;
            slow = slow -> next;
            if(fast == slow) return true;
        }
        return false;
    }
};

File: Algo/Leetcode/141LinkedListCycle.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def hasCycle(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        fast = head
        while fast is not None and fast.next is not None:
            fast = fast.next.next
            head = head.next
            if fast == head:
                return True

        return False